accord-core/src/main/java/accord/topology/TopologyManager.java

```diff
@@ -258,12 +290,45 @@ private Epochs(EpochState[] epochs, List<Notifications> pending, List<FutureEpoc
                 Invariants.requireArgument(futureEpochs.get(i).epoch == futureEpochs.get(i - 1).epoch + 1);
             for (int i = 1; i < epochs.length; i++)
                 Invariants.requireArgument(epochs[i].epoch() == epochs[i - 1].epoch() - 1);
-            this.epochs = epochs;
-        }
+            int truncateFrom = -1;
```

Comment:
Maybe this can be in a specific method, rather than being imposed in the constructor? Would perhaps be clearer to invoke it deliberately, even if we do it in all the same places.